zokrates_core/src/semantics.rs

```diff
@@ -762,24 +763,40 @@ impl<'ast, T: Field> Checker<'ast, T> {
                             true => {}
                         };
 
-                        self.functions.insert(
-                            DeclarationFunctionKey::with_location(
-                                module_id.to_path_buf(),
-                                declaration.id,
-                            )
-                            .signature(funct.signature.clone()),
-                        );
-                        symbols.push(
-                            TypedFunctionSymbolDeclaration::new(
-                                DeclarationFunctionKey::with_location(
-                                    module_id.to_path_buf(),
-                                    declaration.id,
-                                )
-                                .signature(funct.signature.clone()),
-                                TypedFunctionSymbol::Here(funct),
-                            )
-                            .into(),
-                        );
+                        // parameters defined on a non-entrypoint function should not be set as private
+                        match (state.main_id != module_id || declaration.id != "main")
+                            && funct.arguments.iter().any(|a| a.private)
+                        {
+                            true => {
+                                errors.push(
+                                    ErrorInner {
+                                        pos: Some(pos),
+                                        message: "Private argument(s) are only allowed on the entrypoint function".into(),
```

Comment:
This is not strict enough: in `from_ast.rs` we have the following:

```rust
let is_private = param
            .visibility
            .map(|v| match v {
                pest::Visibility::Private(_) => true,
                pest::Visibility::Public(_) => false,
            })
            .unwrap_or(false);
```

Based on this, the semantic checked cannot differentiate between `a` and `public a`, but the latter should not be allowed.

`absy` should be changed to include `Option<bool>` for the visibility, and it should be checked to be `None` for non main, and default to `true` for main.

Reply:
Also the book seems to be lacking documentation on the visibility keywords, and generally on the entry point.